src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,260 @@
+                    self.results[org_id][string]
+                except KeyError:
+                    unmapped_org_strings[org_id].add(string)
+
+        return KeyCollection(unmapped_org_strings)
+
+    def get_mapped_results(self) -> MutableMapping[int, MutableMapping[str, int]]:
+        """
+        Return the results.
+        """
+        return self.results
+
+    def get_mapped_key_strings_to_ints(self) -> MutableMapping[str, int]:
+        """
+        Return the results, but formatted as the following:
+            {
+                "1:a": 10,
+                "1:b": 11,
+                "1:c", 12,
+                "2:e": 13
+            }
+        This is for when we use indexer_cache.set_many()
+        """
+        cache_key_results: MutableMapping[str, int] = {}
+        for org_id, result_dict in self.results.items():
+            for string, id in result_dict.items():
+                key = f"{org_id}:{string}"
+                cache_key_results[key] = id
+
+        return cache_key_results
+
+
+class PGStringIndexerV2(Service):
+    """
+    Provides integer IDs for metric names, tag keys and tag values
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _get_db_records(self, db_keys: KeyCollection) -> Any:
+        conditions = []
+        for pair in db_keys.as_tuples():
+            organization_id, string = pair
+            conditions.append(Q(organization_id=int(organization_id), string=string))
+
+        query_statement = reduce(or_, conditions)
+
+        return StringIndexerTable.objects.filter(query_statement)
+
+    def bulk_record(
+        self, org_strings: MutableMapping[int, Set[str]]
+    ) -> MutableMapping[int, MutableMapping[str, int]]:
+        """
+        TODO(meredith): add doc string
+        """
+        mapped_results: MutableMapping[int, MutableMapping[str, int]] = {}
+
+        cache_keys = KeyCollection(org_strings)
+
+        # Step 1: Lookup ids in the cache and add them to the
+        # cache KeyCollection results.
+        cache_hits = 0
+        cache_misses = 0
+        cache_key_strs = cache_keys.as_strings()
+        cache_results = indexer_cache.get_many(cache_key_strs)
+        for cache_key in cache_key_strs:
+            result = cache_results.get(cache_key)
+            org_id, string = cache_key.split(":")
+            if result:
+                cache_keys.add_key_result((int(org_id), string), result)
+                cache_hits += 1
+            else:
+                cache_misses += 1
+
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "true", "caller": "get_many_ids"},
+            amount=cache_hits,
+        )
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "false", "caller": "get_many_ids"},
+            amount=cache_misses,
+        )
+        mapped_results.update(cache_keys.get_mapped_results())
```

Comment:
Please create `mapped_results` here, not at the beginning of the method. 
If you create a variable immediately before using it the reviewer would not have to keep that in mind till you get to the point where the variable is used reducing the cognitive load.

```
mapped_results = cache_keys.get_mapped_results()
```